functorch/experimental/_map.py

```diff
@@ -320,66 +356,25 @@ def map_fake_tensor_mode(mode, f, num_mapped, *args):
         return map_dense(f, num_mapped, *args)
 
 
-@map_impl.py_impl(DispatchKey.Functionalize)
-def map_func(f, num_mapped, *args):
-    reapply_views = torch._C._functionalization_reapply_views_tls()
+@map_impl.py_functionalize_impl()
+def map_functionalize(ctx, f, num_mapped, *args):
     xs = args[:num_mapped]
     pos_args = args[num_mapped:]
-    unwrapped_xs = _unwrap_all_tensors_from_functional(xs, reapply_views=reapply_views)
-    unwrapped_args = _unwrap_all_tensors_from_functional(
-        pos_args, reapply_views=reapply_views
-    )
-    mode = "mutations_and_views" if reapply_views else "mutations"
+    unwrapped_xs = ctx.unwrap(xs)
+    unwrapped_args = ctx.unwrap(pos_args)
```

Comment:
Should unwrap/wrap unwrap a single Tensor or do the tree_map_only(torch.Tensor)?

Reply:
All 3 of the functionalization rules that we have end up doing a tree_map, so it felt generic enough that I abstracted it away into the `ctx` instead of making the higher order op author spell it out.

But making the higher order op author do it also seems reasonable, let me know if you have a preference

Reply:
No real preference, I agree that everywhere we  use it it seems to be used with a tree_map_only so that sounds reasonable to me